src/ray/raylet/node_manager.cc

```diff
@@ -359,6 +361,38 @@ void NodeManager::HandleActorCreation(const ActorID &actor_id,
   }
 }
 
+void NodeManager::GetActorTasksFromList(const ActorID &actor_id,
+                                        const std::list<Task> &tasks,
+                                        std::unordered_set<TaskID> &tasks_to_remove) {
+  for (auto const &task : tasks) {
+    auto const &spec = task.GetTaskSpecification();
+    if (actor_id == spec.ActorId()) {
+      tasks_to_remove.insert(spec.TaskId());
+    }
+  }
+}
+
+void NodeManager::CleanUpTasksForDeadActor(const ActorID &actor_id) {
+  // TODO(rkn): The code below should be cleaned up when we improve the
+  // SchedulingQueue API.
+  std::unordered_set<TaskID> tasks_to_remove;
+
+  GetActorTasksFromList(actor_id, local_queues_.GetUncreatedActorMethods(),
+                        tasks_to_remove);
+  GetActorTasksFromList(actor_id, local_queues_.GetWaitingTasks(), tasks_to_remove);
+  GetActorTasksFromList(actor_id, local_queues_.GetPlaceableTasks(), tasks_to_remove);
+  GetActorTasksFromList(actor_id, local_queues_.GetReadyTasks(), tasks_to_remove);
+  GetActorTasksFromList(actor_id, local_queues_.GetRunningTasks(), tasks_to_remove);
+  GetActorTasksFromList(actor_id, local_queues_.GetBlockedTasks(), tasks_to_remove);
+
+  auto removed_tasks = local_queues_.RemoveTasks(tasks_to_remove);
+  for (auto const &task : removed_tasks) {
+    const TaskSpecification &spec = task.GetTaskSpecification();
+    TreatTaskAsFailed(spec);
+    task_dependency_manager_.TaskCanceled(spec.TaskId());
```

Comment:
Maybe makes sense to move this line into `TreatTaskAsFailed`?

Reply:
It only makes sense to call `task_dependency_manager_.TaskCanceled` for tasks that have been queued, right?

We call `TreatTaskAsFailed` within `SubmitTask` on tasks that have not been queued yet, and in that case we don't do `task_dependency_manager_.TaskCanceled`, which is why I left it out of `TreatTaskAsFailed`.

Note that this is the same reason that I left out `local_queues_.RemoveTasks` from `TreatTaskAsFailed`.

However, I'm not totally sure about this approach, so if you think there's a cleaner way then we can definitely do that.

Reply:
Ah, got it. I think this is okay then.